llvm/lib/Target/SPIRV/SPIRVModuleAnalysis.cpp

```diff
@@ -1738,15 +1738,42 @@ void addInstrRequirements(const MachineInstr &MI,
     break;
   case SPIRV::OpConvertHandleToImageINTEL:
   case SPIRV::OpConvertHandleToSamplerINTEL:
-  case SPIRV::OpConvertHandleToSampledImageINTEL:
+  case SPIRV::OpConvertHandleToSampledImageINTEL: {
     if (!ST.canUseExtension(SPIRV::Extension::SPV_INTEL_bindless_images))
       report_fatal_error("OpConvertHandleTo[Image/Sampler/SampledImage]INTEL "
                          "instructions require the following SPIR-V extension: "
                          "SPV_INTEL_bindless_images",
                          false);
+    SPIRVGlobalRegistry *GR = ST.getSPIRVGlobalRegistry();
+    SPIRV::AddressingModel::AddressingModel AddrModel;
+    unsigned PointerSize = ST.getPointerSize();
+    AddrModel = PointerSize == 32 ? SPIRV::AddressingModel::Physical32
+                                  : SPIRV::AddressingModel::Physical64;
+    SPIRVType *TyDef = GR->getSPIRVTypeForVReg(MI.getOperand(1).getReg());
+    if (!(TyDef->getOpcode() == SPIRV::OpTypeImage &&
+          MI.getOpcode() == SPIRV::OpConvertHandleToImageINTEL) &&
+        !(TyDef->getOpcode() == SPIRV::OpTypeSampler &&
+          MI.getOpcode() == SPIRV::OpConvertHandleToSamplerINTEL) &&
+        !(TyDef->getOpcode() == SPIRV::OpTypeSampledImage &&
+          MI.getOpcode() == SPIRV::OpConvertHandleToSampledImageINTEL))
+      report_fatal_error("Incorrect return type of the instruction", false);
```

Comment:
I suggest to split this single check into 3: one per instruction, I would make code more readable and also would help to easily include the faulty instruction's name in the error message.